docs/design/environments-implementation.md

```diff
@@ -0,0 +1,193 @@
+   limitations under the License.
+-->
+
+# Overview
+
+Environment profiles (or environment for short) defines a set of libraries and when Docker is being used, a Docker image in order to run an experiment or a notebook. 
+
+Docker and/or VM image (such as AMI: Amazon Machine Images) defines the base layer of the environment. 
+
+On top of that, users can define a set of libraries (such as Python/R) to install, we call it kernel.
+
+**Example of Environment**
+
+```
+
+     +-------------------+
+     |+-----------------+|
+     || Python=3.7      ||
+     || Tensorflow=2.0  ||
+     |+---Exp Dependency+|
+     |+-----------------+|
+     ||OS=Ubuntu16.04   ||
+     ||CUDA=10.2        ||
+     ||GPU_Driver=375.. ||
+     |+---Base Library--+|
+     +-------------------+
+```
+
+As you can see, There're base libraries, such as what OS, CUDA version, GPU driver, etc. They can be achieved by specifying a VM image / Docker image.
+
+On top of that, user can bring their dependencies, such as different version of Python, Tensorflow, Pandas, etc.
+
+**How users use environment?**
+
+Users can save different environment configs which can be also shared across the platform. Environment profiles can be used to run a notebook (e.g. by choosing different kernel from Jupyter), or an experiment. Predefined experiment library includes what environment to use so users don't have to choose which environment to use.
+
+```
+
+        +-------------------+
+        |+-----------------+|       +------------+
+        || Python=3.7      ||       |User1       |
+        || Tensorflow=2.0  ||       +------------+
+        |+---Kernel -------+|       +------------+
+        |+-----------------+|<----+ |User2       |
+        ||OS=Ubuntu16.04   ||     + +------------+
+        ||CUDA=10.2        ||     | +------------+
+        ||GPU_Driver=375.. ||     | |User3       |
+        |+---Base Library--+|     | +------------+
+        +-----Default-Env---+     |
+                                  |
+                                  |
+        +-------------------+     |
+        |+-----------------+|     |
+        || Python=3.3      ||     |
+        || Tensorflow=2.0  ||     |
+        |+---kernel--------+|     |
+        |+-----------------+|     |
+        ||OS=Ubuntu16.04   ||     |
+        ||CUDA=10.3        ||<----+
+        ||GPU_Driver=375.. ||
+        |+---Base Library--+|
+        +-----My-Customized-+
+```
+
+There're two environments in the above graph, "Default-Env" and "My-Customized", which can have different combinations of libraries for different experiments/notebooks. Users can choose different environments for different experiments as they want.
+
+Environments can be added/listed/deleted/selected through CLI/SDK/UI.
+
+# Implementation
+
+## Environment API definition
+
+Let look at what object definition looks like to define an environment, API of environment looks like:
+
+```
+    name: "my_submarine_env",
+    vm-image: "...",
```

Comment:
I do n’t quite understand the meaning of `vm-image`, Does `vm-image` refer to the ec2 machine type on aws for example?
https://amazonaws-china.com/cn/ec2/instance-types/
But the environment of ec2 server cannot be provided by docker image.

Reply:
It is different from the instance-type.  
I just added a note to the same doc: 

> Docker and/or VM-image (such as, VirtualBox/VMWare images, Amazon Machine Images - > AMI, Or custom image of Azure VM) defines the base layer of the environment. Please note > that VM-image is different from VM instance type,